asmcomp/selectgen.ml

```diff
@@ -93,8 +93,9 @@ let size_machtype mty =
 let size_expr (env:environment) exp =
   let rec size localenv = function
       Cconst_int _ | Cconst_natint _ -> Arch.size_int
-    | Cconst_symbol _ | Cconst_pointer _ | Cconst_natpointer _ ->
+    | Cconst_pointer _ | Cconst_natpointer _ ->
         Arch.size_addr
+    | Cconst_symbol (_, ty, _) -> size_machtype ty
```

Comment:
Here and below (in the `Cconst_symbol` case of `emit_expr`) I think we are assuming that `Arch.size_addr` is equal to `Arch.size_int`.  Pointers are still "pointer width" whether or not they are GC-scannable.  It looks like this constraint holds for all backends, at present, but I wonder if it should be enforced somewhere.

This case in `size_expr` should probably cause a fatal error if `ty` is not either `typ_val` or `typ_int`.

Reply:
I don't follow this. Where is this assumption being made?

Reply:
I see where the assumption is now. Cconst_symbol always evaluates to an address.

However size_addr and size_int seem to have always agreed, on every platform, and I think much of Cmm assumes they're equal. Should we consider deleting one? (cc @xavierleroy)

Reply:
I agree with @mshinwell that `Cconst_symbol` must have size `size_addr`, because it evaluates to a pointer.

Cmmgen and all the OCaml front- and middle-end certainly assume that `size_int = size_addr`.  As far as the back-end is concerned, the two sizes must also be equal if types `Int` and `Addr` are mapped to the same class of allocatable registers, which is the case for all our target processors.  (I think it wasn't true for the M68k...)  There might be other places where this is assumed.

So, I have no problem assuming 1-`size_int = size_addr` everywhere, and 2- eventually replace those two parameters by a single parameter.